R/kernel.r

```diff
@@ -101,11 +101,51 @@ handle_shell = function() {
         history(msg)
     } else if (msg$header$msg_type == "complete_request") {
         complete(msg)
+    } else if (msg$header$msg_type == "is_complete_request") {
+        is_complete(msg)
     } else {
         print(c("Got unhandled msg_type:", msg$header$msg_type))
     }
 },
 
+is_complete = function(request) {
+    code <- request$content$code
+    message <- tryCatch({
+        evaluate:::parse_all(code)
+        # the code compiles, so we are complete (either no code at all / only
+        # comments or syntactically correct code)
+        "complete" # no function, so no return!
+    }, warning = function(e){
+        return("warning") # IMO this can't happend...
+    }, error = function(e){
+        return(e$message)
+    })
+    # One of 'complete', 'incomplete', 'invalid', 'unknown'
+    status <- if(message=="complete"){
+        # syntactical complete code
+        "complete"
+    } else if (grepl("unexpected end of input", message)){
+        # missing closing parenthesis
+        "incomplete"
+    } else if (grepl("unexpected INCOMPLETE_STRING", message)){
+        # missing closing quotes
+        "incomplete"
+    } else if(message=="warning"){
+        "unknown"
+    } else {
+        # all else
+        "invalid"
+    }
+    content = list(status=status)
+    if (status == "incomplete"){
+        # we don't try to guess the indention level and just return zero indention
+        # That's fine because R has braces... :-)
+        # TODO: do some guessing?
+        content = c(content, indent = "")
```

Comment:
I generally add things to a list by doing `content$indent = ""`. Both work, I don't know which is considered more idiomatic R.